src/plugins/nestedHeaders/stateManager/index.js

```diff
@@ -0,0 +1,330 @@
+   */
+  setState(nestedHeadersSettings) {
+    this.#sourceSettings.setData(nestedHeadersSettings);
+    let hasError = false;
+
+    try {
+      this.#headersTree.buildTree();
+    } catch (ex) {
+      this.#headersTree.clear();
+      this.#sourceSettings.clear();
+      hasError = true;
+    }
+
+    this.#stateMatrix = matrixGenerator(this.#headersTree.getRoots());
+
+    return hasError;
+  }
+
+  /**
+   * Sets columns limit to the state will be trimmed. All headers (colspans) which
+   * overlap the column limit will be reduced to keep the structure solid.
+   *
+   * @param {number} columnsCount The number of columns to limit to.
+   */
+  setColumnsCountLimit(columnsCount) {
+    this.#sourceSettings.setColumnsCountLimit(columnsCount);
+  }
+
+  /**
+   * Merges settings with current plugin state.
+   *
+   * By default only foreign keys are merged with source state and passed to the tree. But only
+   * known keys are exported to matrix.
+   *
+   * @param {object[]} settings An array of objects to merge with the current source settings.
+   *                            It is a requirement that every object has `row` and `col` properties
+   *                            which points to the specific header settings object.
+   */
+  mergeStateWith(settings) {
+    const transformedSettings = arrayMap(settings, ({ row, ...rest }) => {
+      return {
+        row: row < 0 ? this.rowCoordsToLevel(row) : row,
+        ...rest,
+      };
+    });
+
+    this.#sourceSettings.mergeWith(transformedSettings);
+    this.#headersTree.buildTree();
+    this.#stateMatrix = matrixGenerator(this.#headersTree.getRoots());
+  }
+
+  /**
+   * Maps the current state with a callback. For each header settings the callback function
+   * is called. If the function returns value that value is merged with the state.
+   *
+   * By default only foreign keys are merged with source state and passed to the tree. But only
+   * known keys are exported to matrix.
+   *
+   * @param {Function} callback A function that is called for every header source settings.
+   *                            Each time the callback is called, the returned value extends
+   *                            header settings.
+   */
+  mapState(callback) {
+    this.#sourceSettings.map(callback);
+    this.#headersTree.buildTree();
+    this.#stateMatrix = matrixGenerator(this.#headersTree.getRoots());
+  }
+
+  /**
+   * Maps the current tree nodes with a callback. For each node the callback function
+   * is called. If the function returns value that value is added to returned array.
+   *
+   * @param {Function} callback A function that is called for every tree node.
+   *                            Each time the callback is called, the returned value is
+   *                            added to returned array.
+   * @returns {Array}
+   */
+  mapNodes(callback) {
+    return arrayReduce(this.#headersTree.getRoots(), (acc, rootNode) => {
```

Comment:
The method can return any values (not node instances) based on node data. So either `nodes` and `mappedNodes` do not match here. IMO the `acc` in that not complicated example is a pretty good name. It could be replaced as `mappedValues` or `mappedResults`, but then for me, the meaning is almost the same as `acc`.